test/jobs/test_samtools.py

```diff
@@ -0,0 +1,111 @@
+import re
+import os
+
+from .. import set_config
+from ..factories.alignment_input import create_cram_input
+from ..factories.batch import create_local_batch
+from ..factories.config import PipelineConfig, WorkflowConfig
+from ..factories.sequencing_group import create_sequencing_group
+from .helpers import get_command_str
+
+from cpg_workflows.jobs.samtools import samtools_stats
+from cpg_utils import Path
+from cpg_workflows.filetypes import CramPath
+from cpg_utils.hail_batch import image_path
+
+
+def default_config() -> PipelineConfig:
+    return PipelineConfig(
+        workflow=WorkflowConfig(
+            dataset='samtools-test',
+            access_level='test',
+            sequencing_type='genome',
+            check_inputs=False,
+        ),
+        images={
+            'samtools': 'test_image',
+        },
+        other={
+            'references': {
+                'broad': {
+                    'ref_fasta': 'hg38_reference.fa',
+                    'dragmap_prefix': 'gs://a-cpg-bucket/dragen_reference/',
+                }
+            },
+        },
+    )
+
+
+class TestSamtoolsRun:
+    def test_create_cram_input_method_creates_inputs(self, tmp_path: Path):
+        # ---- Test setup
+        config = default_config()
+        set_config(config, tmp_path / 'config.toml')
+
+        cram_pth = create_cram_input(
+            location=tmp_path, prefix='test', index=True, reference_assembly='GRCh38.fa'
+        )
+        # ---- Check CRAM and CRAI files exist
+        assert os.path.exists(tmp_path / 'test.cram')
+        assert os.path.exists(tmp_path / 'test.cram.crai')
+
+    def test_run_samtools(self, tmp_path: Path):
```

Comment:
This is a good first attempt which is asserting a lot of the things we'd want to test. However, ideally I'd like to see these tests split into smaller tests with much more descriptive names testing one aspect at a time. It should be clearly obvious what a test is doing just by reading the function name as opposed to a general names like `test_run_samtools` and `test_sample_command` in which the intent is not clear. Some examples of how I would structure tests:

test_creates_one_job
test_can_reuse_existing_output_path
test_can_overwrite_existing_output_path
test_sets_job_attrs_or_sets_default_attrs_if_not_supplied
test_uses_samtools_image_specified_in_config
test_uses_reference_in_workflow_config_section_if_set
test_uses_broad_reference_as_default_if_reference_not_set_in_workflow_config_section
test_uses_fail_safe_copy_on_cram_path_and_index_in_bash_command
test_samtools_writes_to_resource_file_named_output_stats
test_batch_writes_samtools_stats_file_to_output_path

You can test for things in the job's bash command using a regex assertion, and the `get_command_str` helper.